hal/src/common/rtc.rs

```diff
@@ -0,0 +1,305 @@
+                2 => w.prescaler().div2(),
+                4 => w.prescaler().div4(),
+                8 => w.prescaler().div8(),
+                16 => w.prescaler().div16(),
+                32 => w.prescaler().div32(),
+                64 => w.prescaler().div64(),
+                128 => w.prescaler().div128(),
+                256 => w.prescaler().div256(),
+                512 => w.prescaler().div512(),
+                1024 => w.prescaler().div1024(),
+                _ => unreachable!(),
+            };
+            // clear timer on match...
+            w.matchclr().set_bit();
+            // and enable RTC.
+            w.enable().set_bit()
+        });
+    }
+
+    fn wait(&mut self) -> nb::Result<(), Void> {
+        if self.mode0().intflag.read().cmp0().bit_is_set() {
+            // Writing a 1 clears the flag
+            self.mode0().intflag.modify(|_, w| w.cmp0().set_bit());
+            Ok(())
+        } else {
+            Err(nb::Error::WouldBlock)
+        }
+    }
+}
+
+impl InterruptDrivenTimer for Rtc {
+    /// Enable the interrupt generation for this hardware timer.
+    /// This method only sets the clock configuration to trigger
+    /// the interrupt; it does not configure the interrupt controller
+    /// or define an interrupt handler.
+    fn enable_interrupt(&mut self) {
+        self.mode0().intenset.write(|w| w.cmp0().set_bit());
+    }
+
+    /// Disables interrupt generation for this hardware timer.
+    /// This method only sets the clock configuration to prevent
+    /// triggering the interrupt; it does not configure the interrupt
+    /// controller.
+    fn disable_interrupt(&mut self) {
+        self.mode0().intenclr.write(|w| w.cmp0().set_bit());
+    }
+}
+
+/// Helper type for computing cycles and divider given frequency
+#[derive(Debug, Clone, Copy)]
+pub struct TimerParams {
+    pub divider: u16,
+    pub cycles: u32,
+}
+
+impl TimerParams {
+    /// calculates RTC timer paramters based on the input frequency-based
+    /// timeout.
+    pub fn new<T>(timeout: T, src_freq: u32) -> Self
+    where
+        T: Into<Hertz>,
+    {
+        let timeout = timeout.into();
+        let ticks: u32 = src_freq / timeout.0.max(1);
+        TimerParams::new_from_ticks(ticks)
+    }
+
+    /// calculates RTC timer paramters based on the input period-based timeout.
+    pub fn new_us<T>(timeout: T, src_freq: u32) -> Self
+    where
+        T: Into<Nanoseconds>,
+    {
+        let timeout = timeout.into();
+        let ticks: u32 = (timeout.0 as u64 * src_freq as u64 / 1_000_000_000_u64) as u32;
+        Self::new_from_ticks(ticks)
+    }
+
+    /// Common helper function that gets the best divider & calculates cycles
+    /// with that divider.
+    fn new_from_ticks(ticks: u32) -> Self {
+        let mut divider = ((ticks >> 16) + 1).next_power_of_two();
+        if divider > 1024 {
+            // lame. Can we catch this at compile time? Probably??
+            divider = 1024;
+        }
```

Comment:
@TDHolmes, here's one thought. In the vein of "parse don't validate", you calculate a valid divider here, but then you have to re-validate it again in the code for `start` above. I think it might make sense to define an `enum` to capture the "parsing" you do here.

Reply:
This makes sense. Trying to directly use the `PRESCALER_A` enum but it isn't quite working. Will pull out the debugger tomorrow to see what's up

Reply:
Figured it out. Was using the enum to divide and calculate ticks when I should have been using the divider value